sqlglot/optimizer/annotate_types.py

```diff
@@ -231,22 +235,45 @@ def annotate_scope(self, scope: Scope) -> None:
             elif isinstance(expression, exp.SetOperation) and len(expression.left.selects) == len(
                 expression.right.selects
             ):
-                if expression.args.get("by_name"):
-                    r_type_by_select = {s.alias_or_name: s.type for s in expression.right.selects}
-                    selects[name] = {
-                        s.alias_or_name: self._maybe_coerce(
-                            t.cast(exp.DataType, s.type),
-                            r_type_by_select.get(s.alias_or_name) or exp.DataType.Type.UNKNOWN,
-                        )
-                        for s in expression.left.selects
-                    }
-                else:
-                    selects[name] = {
-                        ls.alias_or_name: self._maybe_coerce(
-                            t.cast(exp.DataType, ls.type), t.cast(exp.DataType, rs.type)
-                        )
-                        for ls, rs in zip(expression.left.selects, expression.right.selects)
-                    }
+                col_types: t.Dict[str, exp.DataType.Type] = self._setop_cols.get(id(expression), {})
```

Comment:
Why do we need the annotation here? Also, we can do this instead:

```python
selects[name] = column_types = self._setop_column_types.setdefault(id(expression), {})
```

Then we'll be able to remove lines 274-276, because both dicts will be updated due to mutation.

Reply:
That annotation was explicitly requested by mypy, probably due to the dict operations.